tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,217 @@
+{
+    const auto fd = args[0].as<uint32_t>();
+    const auto iov_ptr = args[1].as<uint32_t>();
+    const auto iov_cnt = args[2].as<uint32_t>();
+    const auto nwritten_ptr = args[3].as<uint32_t>();
+
+    std::vector<uvwasi_ciovec_t> iovs(iov_cnt);
+    // TODO: not sure what to pass as end, passing memory size...
+    uvwasi_errno_t ret = uvwasi_serdes_readv_ciovec_t(instance.memory->data(),
+        instance.memory->size(), iov_ptr, iovs.data(), static_cast<uvwasi_size_t>(iovs.size()));
+    if (ret != UVWASI_ESUCCESS)
+        return fizzy::Value{uint32_t{ret}};
+
+    uvwasi_size_t nwritten;
+    ret = uvwasi_fd_write(&state, static_cast<uvwasi_fd_t>(fd), iovs.data(),
+        static_cast<uvwasi_size_t>(iovs.size()), &nwritten);
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), nwritten_ptr, nwritten);
+
+    return fizzy::Value{uint32_t{ret}};
+}
+
+fizzy::ExecutionResult wasi_fd_read(
+    fizzy::Instance& instance, fizzy::span<const fizzy::Value> args, int)
+{
+    const auto fd = args[0].as<uint32_t>();
+    const auto iov_ptr = args[1].as<uint32_t>();
+    const auto iov_cnt = args[2].as<uint32_t>();
+    const auto nread_ptr = args[3].as<uint32_t>();
+
+    std::vector<uvwasi_iovec_t> iovs(iov_cnt);
+    // TODO: not sure what to pass as end, passing memory size...
+    uvwasi_errno_t ret = uvwasi_serdes_readv_iovec_t(instance.memory->data(),
+        instance.memory->size(), iov_ptr, iovs.data(), static_cast<uvwasi_size_t>(iovs.size()));
+    if (ret != UVWASI_ESUCCESS)
+        return fizzy::Value{uint32_t{ret}};
+
+    uvwasi_size_t nread;
+    ret = uvwasi_fd_read(&state, static_cast<uvwasi_fd_t>(fd), iovs.data(),
+        static_cast<uvwasi_size_t>(iovs.size()), &nread);
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), nread_ptr, nread);
+
+    return fizzy::Value{uint32_t{ret}};
+}
+
+fizzy::ExecutionResult wasi_fd_prestat_get(
+    fizzy::Instance& instance, fizzy::span<const fizzy::Value> args, int)
+{
+    const auto fd = args[0].as<uint32_t>();
+    const auto prestat_ptr = args[1].as<uint32_t>();
+
+    uvwasi_prestat_t buf;
+    uvwasi_errno_t ret = uvwasi_fd_prestat_get(&state, fd, &buf);
+
+    uvwasi_serdes_write_prestat_t(instance.memory->data(), prestat_ptr, &buf);
+
+    return fizzy::Value{uint32_t{ret}};
+}
+
+fizzy::ExecutionResult wasi_environ_sizes_get(
+    fizzy::Instance& instance, fizzy::span<const fizzy::Value> args, int)
+{
+    const auto environc = args[0].as<uint32_t>();
+    const auto environ_buf_size = args[1].as<uint32_t>();
+    // TODO: implement properly (only returns 0 now)
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), environc, 0);
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), environ_buf_size, 0);
+    return fizzy::Value{uint32_t{UVWASI_ESUCCESS}};
+}
+
+bool run(int argc, const char** argv)
+{
+    // TODO: support both wasi_snapshot_preview1 and wasi_unstable?
+    const std::vector<fizzy::ImportedFunction> wasi_functions = {
```

Comment:
```suggestion
    const fizzy::ImportedFunction wasi_functions[] = {
```

Reply:
```
/home/builder/project/tools/wasi/wasi.cpp: In function 'bool {anonymous}::run(int, const char**)':
/home/builder/project/tools/wasi/wasi.cpp:160:62: error: could not convert '(const fizzy::ImportedFunction*)(& wasi_functions)' from 'const fizzy::ImportedFunction*' to 'std::vector<fizzy::ImportedFunction>'
  160 |     auto imports = fizzy::resolve_imported_functions(module, wasi_functions);
      |                                                              ^~~~~~~~~~~~~~
      |                                                              |
      |                                                              const fizzy::ImportedFunction*
```

`resolve` requires a vector currently, I'd leave this do be fixed after this is merged.